bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/Bookie.java

```diff
@@ -16,154 +15,80 @@
+import org.apache.bookkeeper.meta.LedgerManagerFactory;
+import org.apache.bookkeeper.common.util.Watcher;
+import java.util.PrimitiveIterator;
 
-    public static final long METAENTRY_ID_LEDGER_KEY = -0x1000;
-    public static final long METAENTRY_ID_FENCE_KEY  = -0x2000;
-    public static final long METAENTRY_ID_FORCE_LEDGER  = -0x4000;
-    static final long METAENTRY_ID_LEDGER_EXPLICITLAC  = -0x8000;
+public interface Bookie {
 
-    private final LedgerDirsManager ledgerDirsManager;
-    protected final Supplier<BookieServiceInfo> bookieServiceInfoProvider;
-    private final LedgerDirsManager indexDirsManager;
-    LedgerDirsMonitor dirsMonitor;
+    void start();
+    void join() throws InterruptedException;
+    boolean isRunning();
+    int getExitCode();
+    int shutdown();
 
-    // Registration Manager for managing registration
-    protected final MetadataBookieDriver metadataDriver;
+    boolean isAvailableForHighPriorityWrites();
+    boolean isReadOnly();
 
-    private int exitCode = ExitCode.OK;
+    // TODO: replace callback with futures
+    // TODO: replace ackBeforeSync with flags
+    void addEntry(ByteBuf entry, boolean ackBeforeSync, WriteCallback cb, Object ctx, byte[] masterKey)
+            throws IOException, BookieException, InterruptedException;
+    void recoveryAddEntry(ByteBuf entry, WriteCallback cb, Object ctx, byte[] masterKey)
+            throws IOException, BookieException, InterruptedException;
+    void forceLedger(long ledgerId, WriteCallback cb, Object ctx);
+    void setExplicitLac(ByteBuf entry, WriteCallback writeCallback, Object ctx, byte[] masterKey)
+            throws IOException, InterruptedException, BookieException;
+    ByteBuf getExplicitLac(long ledgerId) throws IOException, NoLedgerException;
 
-    private final ConcurrentLongHashMap<byte[]> masterKeyCache = new ConcurrentLongHashMap<>();
+    // these can probably be moved out and called directly on ledgerdirmanager
+    long getTotalDiskSpace() throws IOException;
+    long getTotalFreeSpace() throws IOException;
 
-    protected StateManager stateManager;
+    // TODO: Shouldn't this be async?
+    ByteBuf readEntry(long ledgerId, long entryId)
+            throws IOException, NoLedgerException;
+    long readLastAddConfirmed(long ledgerId) throws IOException;
+    PrimitiveIterator.OfLong getListOfEntriesOfLedger(long ledgerId) throws IOException, NoLedgerException;
 
-    // Expose Stats
-    final StatsLogger statsLogger;
-    private final BookieStats bookieStats;
+    /**
+     * Fences a ledger. From this point on, clients will be unable to
+     * write to this ledger. Only recoveryAddEntry will be
+     * able to add entries to the ledger.
+     * This method is idempotent. Once a ledger is fenced, it can
+     * never be unfenced. Fencing a fenced ledger has no effect.
+     */
+    // TODO: replace settable future with completablefuture
+    SettableFuture<Boolean> fenceLedger(long ledgerId, byte[] masterKey)
```

Comment:
yes, that exactly what I meant.

is these TODO to be applied in this patch ?
otherwise I would prefer to not leave TODOs in the code
or at least to create github issues in order to track the work and have references

Reply:
Done.